claude-instructions/.claude/skills/charmcraft/references/charmcraft-yaml-reference.md

```diff
@@ -0,0 +1,540 @@
+    version: "0"
+  - lib: observability_libs.v0.kubernetes_service_patch
+    version: "0"
+
+assumes:
+  - juju >= 3.1
+  - k8s-api
+
+containers:
+  postgresql:
+    resource: postgresql-image
+
+resources:
+  postgresql-image:
+    type: oci-image
+    description: "PostgreSQL container image"
+
+config:
+  options:
+    profile:
+      type: string
+      description: "Resource allocation profile (testing/production)"
+      default: "production"
+
+provides:
+  database:
+    interface: postgresql_client
+
+  metrics-endpoint:
+    interface: prometheus_scrape
+
+requires:
+  certificates:
+    interface: tls-certificates
+    optional: true
+
+  s3-parameters:
+    interface: s3
+    optional: true
+    limit: 1
+
+peers:
+  database-peers:
+    interface: postgresql_peers
+
+storage:
+  pgdata:
+    type: filesystem
+    location: /var/lib/postgresql/data
+    minimum-size: 5G
+
+actions:
+  get-primary:
+    description: "Get the primary database unit"
+
+  create-backup:
+    description: "Create a database backup"
+    params:
+      s3-path:
+        type: string
+        description: "S3 path for backup"
+    required: [s3-path]
+
+links:
+  documentation: https://discourse.charmhub.io/t/postgresql-k8s-docs/9308
+  issues: https://github.com/canonical/postgresql-k8s-operator/issues
+  source: https://github.com/canonical/postgresql-k8s-operator
+  website: https://postgresql.org
+```
+
+## Validation
+
+Charmcraft validates `charmcraft.yaml` automatically during:
+- `charmcraft pack`
+- `charmcraft analyse`
+- `charmcraft init`
+
+Common validation errors:
+- Missing required fields
+- Invalid base configuration
+- Malformed YAML syntax
+- Invalid charm name format
+- Missing parts configuration
+
+## Migration from metadata.yaml
+
+Older charms used separate files:
+- `metadata.yaml` - charm metadata
+- `config.yaml` - configuration options
+- `actions.yaml` - action definitions
+
+Modern charms consolidate everything into `charmcraft.yaml`.
+
+To migrate:
+1. Create `charmcraft.yaml` with `name`, `type`, `title`, `summary`
+2. Copy base/requires/provides/peers from `metadata.yaml`
+3. Copy config options from `config.yaml` to `config:` section
+4. Copy actions from `actions.yaml` to `actions:` section
+5. Add `parts:` section with charm plugin
+6. Test with `charmcraft pack`
+7. Remove old files once working
+
+## Best Practices
+
+1. **Always include `optional: true`** for `requires` relations
+2. **Use semantic naming** for relations (database, not db)
+3. **Provide good descriptions** for all config/actions
+4. **Include links** to documentation and source
+5. **Use charm-libs** for common integrations
+6. **Keep description comprehensive** - it's your Charmhub page
+7. **Test multiple bases** if supporting them
+8. **Version lock libraries** when stable (use major.minor not just major)
+9. **Document breaking changes** in charm description
+10. **Use assumes** to prevent deployment on incompatible Juju
```

Comment:
```suggestion
10. **Use assumes** to prevent deployment on incompatible Juju
11. Always include `additionalProperties: false` in action definitions
```